Use internal HD chain for change outputs (hd split)

jonasschnelli opened the pull request:
#### Main changes
This pull request will result in using `m/0'/1'/k'` for internal keys (change outputs) while keeping `m/0'/0'/k'` for external keys (`getnewaddress`).

* There is still a single keypool. All keypool elements (`CKeyPool`) have now a flag (`fInternal`).
* ~~The keypool will be filled with 80% external keys (round ceil).~~
* The keypool will be filled with 100% external keys + 20% (round ceil) internal keys.
* `getwalletinfo` additionally reports `keypoolsize_hd_internal` (amount of internal key in the 

#### Compatibility:
* This hd split change will only affect new wallets. Current 0.13 HD wallets will continue to only use the external chain.
* Using a 0.14 wallet in <0.14 is not possible (hd chain split requires 0.14)
* Using a 0.13 wallet in 0.14 will result in continue using only the external chain (only `m/0'/0'/k'`).

This change also fixes the keypool +1 offset.

laanwj commented:
Interesting assertion error in CWallet::CreateRawTransaction
```
stderr:
bitcoind: ../../src/wallet/wallet.cpp:2395: bool CWallet::CreateTransaction(const std::vector<CRecipient>&, CWalletTx&, CReserveKey&, CAmount&, int&, std::string&, const CCoinControl*, bool): Assertion `ret' failed.
  File "/home/travis/build/bitcoin/bitcoin/qa/rpc-tests/test_framework/test_framework.py", line 145, in main
    self.run_test()
  File "/home/travis/build/bitcoin/bitcoin/build/../qa/rpc-tests/fundrawtransaction.py", line 498, in run_test
    fundedTx = self.nodes[1].fundrawtransaction(rawTx)
  File "/home/travis/build/bitcoin/bitcoin/qa/rpc-tests/test_framework/coverage.py", line 49, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/travis/build/bitcoin/bitcoin/qa/rpc-tests/test_framework/authproxy.py", line 150, in __call__
    response = self._request('POST', self.__url.path, postdata.encode('utf-8'))
  File "/home/travis/build/bitcoin/bitcoin/qa/rpc-tests/test_framework/authproxy.py", line 130, in _request
    self.__conn.request(method, path, postdata, headers)
  File "/usr/lib/python3.4/http/client.py", line 1125, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python3.4/http/client.py", line 1163, in _send_request
    self.endheaders(body)
  File "/usr/lib/python3.4/http/client.py", line 1121, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.4/http/client.py", line 951, in _send_output
    self.send(msg)
  File "/usr/lib/python3.4/http/client.py", line 886, in send
    self.connect()
  File "/usr/lib/python3.4/http/client.py", line 863, in connect
    self.timeout, self.source_address)
  File "/usr/lib/python3.4/socket.py", line 512, in create_connection
    raise err
  File "/usr/lib/python3.4/socket.py", line 503, in create_connection
    sock.connect(sa)
```

jonasschnelli commented:
Travis is failing because it did what it is supposed to: finding bugs.
Once #9295 is merged, travis should succeed.

paveljanik commented on qa/rpc-tests/keypool.py:
```diff
@@ -28,30 +28,39 @@ def run_test(self):
         try:
             addr = nodes[0].getrawchangeaddress()
-            raise AssertionError('Keypool should be exhausted after three addresses')
+            raise AssertionError('Keypool should be exhausted after three internal addresses (20%)')
```
Still three?

paveljanik commented on qa/rpc-tests/keypool.py:
```diff
@@ -28,30 +28,39 @@ def run_test(self):
+        # the next one should fail
+        try:
+            addr = nodes[0].getnewaddress()
+            raise AssertionError('Keypool should be exhausted after three external addresses (ceil rounded 80% of six)')
```
three?

instagibbs replied:
also, this is 100%, not 80% of 6. 100% vs 20%, not 80% vs 20%.

gmaxwell commented:
> Once #9295 is merged, travis should succeed.

Shall we find out?

jonasschnelli commented:
Rebased.

jonasschnelli commented:
Travis succeeds now.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -140,19 +140,22 @@ void CWallet::DeriveNewChildKey(CKeyMetadata& metadata, CKey& secret)
-        externalChainChildKey.Derive(childKey, hdChain.nExternalChainCounter | BIP32_HARDENED_KEY_LIMIT);
-        metadata.hdKeypath = "m/0'/0'/" + std::to_string(hdChain.nExternalChainCounter) + "'";
+        externalChainChildKey.Derive(childKey, (internal ? hdChain.nInternalChainCounter : hdChain.nExternalChainCounter) | BIP32_HARDENED_KEY_LIMIT);
+        metadata.hdKeypath = "m/0'/" + std::string(internal ? "1'/"+ std::to_string(hdChain.nInternalChainCounter) : "0'/" + std::to_string(hdChain.nExternalChainCounter)) + "'";
```
Seems like it would be better to just access the specific counter once with a post-increment, and use variables here.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2815,30 +2831,42 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
 
-        while (setKeyPool.size() < (nTargetSize + 1))
+        // count amount of available keys (internal, external)
+        // try to generate 80% external keys
```
IMO this is unintuitive and can lead to loss. If users set a keypool size of 100, they expect to be able to safely generate 100 addresses between backups. I usually round down to 90 in advice to give some breathing room, but 80% here will break even that.

Therefore, the full keypool size should be ensured on both chains.

jonasschnelli replied:
Yes. This is a discussion point and I'd like to get others feedback (maybe @gmaxwell and @sipa).
The question is, if you want a keypool of 100 external + 50 (TBD)  internals = total 150 keys, if you set the keypool to 100.

IMO user given values should be respected. If the user chooses keypool=100, the keypool should contain 100 keys. But as said, no strong opinion on that.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -87,7 +87,7 @@ const CWalletTx* CWallet::GetWalletTx(const uint256& hash) const
 }
 
-CPubKey CWallet::GenerateNewKey()
+CPubKey CWallet::GenerateNewKey(bool internal)
```
IMO it would be better to use an enum for internal vs external rather than a bool.

jonasschnelli replied:
> IMO it would be better to use an enum for internal vs external rather than a bool.

Why? Either its internal or external. Once we have more configurable HD key derivation (ext-pubkey-derivation, hd-multisig) we could switch to a enum or a more flexible design.

luke-jr replied:
To avoid implicit casting when/if the type changes for this parameter. Also makes it clearer at the call-locations what the parameter is specifying.

luke-jr reviewed (changes requested):
I don't see code for upgrading existing HD wallets.

Does this implement a BIP that should be mentioned?

jonasschnelli commented:
Overhauled after recommendation from @sipa and @luke-jr.
* The amount of external pre-generated keys now back at 100% from `-keypool=(default 100)`.
* The amount of internal pre-generated keys is +20% from `-keypool=(default 100)`.
* Pre-generate two internal keys at minimum

... this now results in always have 120% keys pre-generated (20% internal key)

jonasschnelli commented:
Updated with a small nit-fix (https://github.com/bitcoin/bitcoin/pull/9294#discussion_r93689981) and a fix for wallet-dump.py RPC test. Travis should succeed now.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2300,17 +2301,20 @@ UniValue getwalletinfo(const JSONRPCRequest& request)
+    obj.push_back(Pair("keypoolsize", (int64_t)kpExternalSize));
+    CKeyID masterKeyID = pwalletMain->GetHDChain().masterKeyID;
+    if (!masterKeyID.IsNull())
+        obj.push_back(Pair("keypoolsize_hd_internal",   (int64_t)(pwalletMain->GetKeyPoolSize() - kpExternalSize)));
```
Might make sense to explicitly count both types and use those counts if we foresee a future with more possible types rather than implicitly count `total - external`. You'd know better than me how likely that is.

jonasschnelli replied:
I think it make sense to keep this as it is: `keypoolsize` should be alined with the `-keypool` conf. That's why `keypoolsize` in `getwalletinfo` responses only the external key-count in the keypool (`-keypool` conf arg also defines the external key count).

instagibbs replied:
Sorry, I agree with what you're saying I just wasn't clear enough. I suppose it's along the lines of @luke-jr 's comment about enum, keeping flexible for the future. Not a blocker, just a suggestion. That can be deferred.

jonasschnelli replied:
Right. This is generally a good idea. IMO the next possible HD extensions (far away) are maybe xpub watch-only-wallets, flexible keypath, hd-multisig.
I can't imagine any change the would require a third (or more then two) type(s) during key derivation as well as in the keypool.

luke-jr replied:
`"keypoolsizes": {"internal": N, "external": N}` sounds like a good idea here.

jonasschnelli replied:
I though about that, but this would break the API while the current PRs change does not.

instagibbs reviewed (commented):
You might also want to have a test that ensures the `ceil` behavior is correct for the 100vs20 split as I think the tests only check evenly divisible or the min value of 2.

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -3129,7 +3151,11 @@ bool CReserveKey::GetReservedKey(CPubKey& pubkey, bool internal)
             vchPubKey = keypool.vchPubKey;
         else {
-            return false;
+            if (pwallet->IsLocked())
```
Right. This is no longer required because we always derive at minimum 2 internal keys. And `ReserveKeyFromKeyPool` will topup the keypool.
Will remove that part.

jonasschnelli committed:
[Wallet] split the keypool in an internal and external part

jonasschnelli commented on src/wallet/wallet.cpp:
```diff
@@ -2809,33 +2831,33 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
+        // generate +20% internal keys (minimum 2 keys)
+        int64_t amountExternal = KeypoolCountExternalKeys();
+        int64_t amountInternal = setKeyPool.size() - amountExternal;
+        int64_t targetInternal = max((int64_t)ceil(nTargetSize * 0.2), (int64_t) 2);
```
> I find it a bit confusing that this value goes from 0(non-existant) to -keypool then to -keypool*.2. Perhaps have it start at 0 on previous commit, then set it to non-0 here (unless there is reasoning).

Fixed. I decided to squash the two commit into a single one because the current PR is a non-splittable change.

TheBlueMatt replied:
Can we just set the target for both internal and external to nTargetSize? It seems super non-user-friendly that the setting which used to mean "can create this many keys/transactions" now means "can create this many keys, or 1/5th as many transactions". I dont think we care all that much about the performance hit, do we?

jonasschnelli replied:
Some weeks back we discusses that and we came up with something that we should create less internal keys. But I'm fine with 100%/100%. Any objections?

TheBlueMatt replied:
I'd be fine with reducing the internal keypool significantly once we have rescan logic that can regenerate keys, but until then I think we absolutely need to have it be 100%.

jonasschnelli commented:
Fixed @instagibbs points. Squashed into a single commit.

instagibbs commented:
utACK 4065d5fd44b9ce68a688c82353822a1cf49efe53

jonasschnelli commented:
Rebased.

luke-jr commented on src/wallet/walletdb.h:
```diff
@@ -58,13 +61,16 @@ class CHDChain
     {
         READWRITE(this->nVersion);
         READWRITE(nExternalChainCounter);
+        if (this->nVersion >= VERSION_HD_CHAIN_SPLIT)
```
This should set nInternalChainCounter for old versions - maybe use `std::numeric_limits<uint32_t>::max()` (and make it private, with all accessors throwing if it's not a sufficient version)?

Also, braces, please.

jonasschnelli replied:
Where do you see the need to set the `nInternalChainCounter` for old versions here? Old Versions will always have the `nInternalChainCounter` set the `0` over `SetNull()` during the constructor call.

A new chain (which then will enable `VERSION_HD_CHAIN_SPLIT`) can only be created when a new HD master key will be set which again will enable `FEATURE_HD_SPLIT`.

luke-jr replied:
It should be setup to prevent accidental usage. If any code were to attempt to use it without a wallet that supports it, we want to throw an error/exception, not silently generate keys we may not be able to recover.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -100,7 +100,7 @@ CPubKey CWallet::GenerateNewKey()
     // use HD key derivation if HD was enabled during wallet creation
     if (IsHDEnabled()) {
-        DeriveNewChildKey(metadata, secret);
+        DeriveNewChildKey(metadata, secret, (CanSupportFeature(FEATURE_HD_SPLIT) ? internal : false));
```
Is there a reason to do this check here, rather than inside `DeriveNewChildKey`?

jonasschnelli replied:
A design question, but IMO `DeriveNewChildKey` should execute the derivation and not care about wallet versions and supported features.

luke-jr replied:
Maybe move it to `private:` then?

TheBlueMatt replied:
Yea, really should be private, I think.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -140,19 +140,22 @@ void CWallet::DeriveNewChildKey(CKeyMetadata& metadata, CKey& secret)
         // example: 1 | BIP32_HARDENED_KEY_LIMIT == 0x80000001 == 2147483649
-        externalChainChildKey.Derive(childKey, hdChain.nExternalChainCounter | BIP32_HARDENED_KEY_LIMIT);
-        metadata.hdKeypath = "m/0'/0'/" + std::to_string(hdChain.nExternalChainCounter) + "'";
+        chainChildKey.Derive(childKey, (internal ? hdChain.nInternalChainCounter : hdChain.nExternalChainCounter) | BIP32_HARDENED_KEY_LIMIT);
```
Seems like this could be significantly simplified with:

```C++
uint32_t &nChainCounter = (internal ? hdChain.nInternalChainCounter : hdChain.nExternalChainCounter);
```

theuni replied:
Since almost everything here is different depending on internal/external, I think it'd be clearer to just make it ```if (internal)``` and separate the behaviors

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -1294,8 +1297,8 @@ bool CWallet::SetHDMasterKey(const CPubKey& pubkey)
-    // ensure this wallet.dat can only be opened by clients supporting HD
-    SetMinVersion(FEATURE_HD);
+    // ensure this wallet.dat can only be opened by clients supporting HD with chain split
+    SetMinVersion(FEATURE_HD_SPLIT);
```
I think this won't work if the user specifies a HD-but-not-split wallet version for `-walletupgrade`?

jonasschnelli replied:
I think we should no longer allow non split HD versions.

TheBlueMatt replied:
This also forcibly upgrades the user's wallet (if from pre-split HD version) if they encrypt their wallet, which I think is unaccptable.

TheBlueMatt replied:
Note that if you change this need to set the newHdChain's version to the appropriate value prior to SetHDChain.

jonasschnelli replied:
We could... one question is: if the users encrypt his 0.13 wallet (HD enabled but no chain split) in 0.14. Do we want to keep the non-hd split or do we want to enforce an upgrade to HD_Split. The later seems preferable from the users perspective (Is there a reason to not split the HD chains if we can?).

TheBlueMatt replied:
I believe it is policy to not upgrade a user's wallet unless they have specifically requested we do so, so, no, I dont think we should use HD_Split at that point.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2870,18 +2873,31 @@ bool CWallet::NewKeyPool()
-            setKeyPool.insert(nIndex);
-        }
-        LogPrintf("CWallet::NewKeyPool wrote %d new keys\n", nKeys);
+        TopUpKeyPool();
```
Probably should check the return value here. Since `TopUpKeyPool` also checks `IsLocked`, perhaps replace it above?

jonasschnelli replied:
The program can never reach this point in `IsLocked()` state because we return at L2871.

luke-jr replied:
`TopUpKeyPool` could return `false` for other reasons as well. There's no need to check `IsLocked` at L2871; it's redundant. So just remove that check and do:

```C++
if (!TopUpKeyPool()) {
    return false;
}
```

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2870,18 +2873,31 @@ bool CWallet::NewKeyPool()
     return true;
 }
 
+size_t CWallet::KeypoolCountExternalKeys()
```
Maybe `KeypoolCountKeys(bool internal)`?

jonasschnelli replied:
Again a design question. setKeyPool.size() - KeypoolCountExternalKeys() = internal key count.

luke-jr replied:
That's going to break as soon as we have more than internal/external key pools...

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2870,18 +2873,31 @@ bool CWallet::NewKeyPool()
+
+    // count amount of external keys
+    size_t amountE = 0;
+    for(const int64_t& id : setKeyPool)
```
Iterating over the entire keypool, even if small (which is not guaranteed), seems excessive to do for every keypool top-up... *especially* if we're reading the wallet db (ie, I/O) for each key.

jonasschnelli replied:
I think we have no other reliable way here and the performance impact should be little. BDB should be pretty fast for that operation and we are only opening the DB once.

luke-jr replied:
Is there some reason not to have a vector or map of separate setKeyPools? (With the current setKeyPool a special case for external keys)

TheBlueMatt replied:
Additionally, this is called at the entry to getwalletinfo, even with HD disabled. While I agree with @luke-jr that there should probably be two separate setKeyPools, at a minimum this function needs an if (!IsHDEnabled()) return false; to avoid causing massive slowdown for folks who are using non-hd wallets with massive keypools (to avoid constant backups).

jonasschnelli replied:
Okay. This is fixed now (https://github.com/bitcoin/bitcoin/pull/9294/files#diff-b2bb174788c7409b671c46ccc86034bdR2886).

luke-jr replied:
That commit merely implements a fix for non-split-HD wallets, not for new wallets...

TheBlueMatt replied:
See comments later in the PR, the performance measurements that @jonasschnelli did indicate there is little to no issue for most uses...if we found out this is not true later, we can fix it as a regression-fix.

luke-jr replied:
Okay, I expected changing this would be a wallet file format change, but upon further investigation I see it is not.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2890,30 +2906,45 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
+        int64_t amountExternal = KeypoolCountExternalKeys();
+        int64_t amountInternal = setKeyPool.size() - amountExternal;
+        int64_t targetInternal = max((int64_t)ceil(nTargetSize * 0.2), (int64_t) 2);
+        int64_t missingExternal = max( (int64_t)(nTargetSize - amountExternal), (int64_t) 0);
```
Shouldn't the cast be directly on `nTargetSize`, rather than the result of the subtraction?

TheBlueMatt replied:
Also, I think you need to max(..., 1) - otherwise some assumptions elsewhere about there always being at least 1 key in keypool after TopUpKeyPool might break.

TheBlueMatt replied:
This should be max(max(nTargetSize, 1) - amountExternal, 0) to avoid violating some of the assumptions about at least one keypool entry being available after TopUpKeyPool returns.

TheBlueMatt replied:
I believe this is a bug, which would need fixing prior to merge.

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2890,30 +2906,45 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
+
+        if (!IsHDEnabled() || !CanSupportFeature(FEATURE_HD_SPLIT))
+        {
+            // don't flag keypool keys internal if we don't use HD
```
Also don't create extra internal keys (already current code behaviour, but comment could be clearer).

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2890,30 +2906,45 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
+        }
+        bool internal = true;
+        CWalletDB walletdb(strWalletFile);
+        for(unsigned int i = 0; i < missingInternal + missingExternal; i++)
```
Better to count down and avoid adding every iteration:

```C++
for (int64_t i = missingInternal + missingExternal; i--; )
```

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2890,30 +2906,45 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
+        for(unsigned int i = 0; i < missingInternal + missingExternal; i++)
         {
             int64_t nEnd = 1;
+            internal = (i < missingInternal);
```
(Assuming counting down, and `internal` initialised to `false` above:)

```C++
if (i < missingInternal) {
    internal = true;
}
```

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2929,14 +2960,24 @@ void CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool)
+            if (!walletdb.ReadPool(id, tmpKeypool))
+                throw runtime_error(std::string(__func__) + ": read failed");
+            if (!HaveKey(tmpKeypool.vchPubKey.GetID()))
+                throw runtime_error(std::string(__func__) + ": unknown key in key pool");
```
These exceptions used to be thrown after the key was removed from the set, allowing a retry to succeed. That is no longer the case here, potentially leading to irrecoverable situations. I'm unsure if this is good or bad.

jonasschnelli replied:
Yes. There are arguments for and against throwing before or after removing. If we throw here, you already ran into (more) significant issues.

luke-jr commented on src/wallet/wallet.h:
```diff
@@ -111,6 +116,13 @@ class CKeyPool
+        else
+        {
+            if (ser_action.ForRead())
+                fInternal = false;
```
Maybe throw an exception if `!ser_action.ForRead() && fInternal`?

Either way, please add braces.

luke-jr commented on src/wallet/wallet.h:
```diff
@@ -712,8 +724,8 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
-    CPubKey GenerateNewKey();
-    void DeriveNewChildKey(CKeyMetadata& metadata, CKey& secret);
+    CPubKey GenerateNewKey(bool internal = false);
+    void DeriveNewChildKey(CKeyMetadata& metadata, CKey& secret, bool internal = false);
```
Do we want `internal` to have a default? Should it be `false`? IIRC currently new keys default to change unless added to the address book, so this default seems to contradict the status quo.

jonasschnelli replied:
I have set the default to false to keep the exiting behaviour.

luke-jr replied:
My point is that the existing behaviour is closer to `true` than `false`.

luke-jr commented on src/wallet/walletdb.h:
```diff
@@ -46,9 +46,12 @@ class CHDChain
     CKeyID masterKeyID; //!< master key hash160
 
-    static const int CURRENT_VERSION = 1;
+    static const int VERSION_HD_BASE        = 1;
```
Missing logic to initialise this version.

jonasschnelli replied:
`VERSION_HD_BASE` is only there to identify versions that do not support HD SPLIT (first HD version introduced in 0.13).

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2315,17 +2316,20 @@ UniValue getwalletinfo(const JSONRPCRequest& request)
     obj.push_back(Pair("txcount",       (int)pwalletMain->mapWallet.size()));
     obj.push_back(Pair("keypoololdest", pwalletMain->GetOldestKeyPoolTime()));
-    obj.push_back(Pair("keypoolsize",   (int)pwalletMain->GetKeyPoolSize()));
+    obj.push_back(Pair("keypoolsize", (int64_t)kpExternalSize));
```
`keypoolsize` should probably be the lower of the two and deprecated. There is no reason to assume prior use didn't look at it for info on change keys.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2315,17 +2316,20 @@ UniValue getwalletinfo(const JSONRPCRequest& request)
-    obj.push_back(Pair("keypoolsize",   (int)pwalletMain->GetKeyPoolSize()));
+    obj.push_back(Pair("keypoolsize", (int64_t)kpExternalSize));
+    CKeyID masterKeyID = pwalletMain->GetHDChain().masterKeyID;
+    if (!masterKeyID.IsNull())
```
Shouldn't this use `CanSupportFeature(FEATURE_HD_SPLIT)`?

jonasschnelli replied:
I think always reporting `keypoolsize_hd_internal` (with 0 in non HD-SPLIT case) can make parsing simpler.

luke-jr replied:
Then always report it? Right now it's conditional on something more or less irrelevant.

But I'm not sure it makes sense to ever report 0 when there are indeed keys that will be used for internal outputs...

jonasschnelli commented:
Added a commit that addresses some of @luke-jr points.
Maybe we can try to not bike-shed this with to many design and code-style nits (braces, etc.) otherwise we will very likely delay this for 0.15.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2296,16 +2296,17 @@ UniValue getwalletinfo(const JSONRPCRequest& request)
+            "  \"txcount\": xxxxxxx,              (numeric) the total number of transactions in the wallet\n"
+            "  \"keypoololdest\": xxxxxx,         (numeric) the timestamp (seconds since Unix epoch) of the oldest pre-generated key in the key pool\n"
+            "  \"keypoolsize\": xxxx,             (numeric) how many new keys are pre-generated\n"
+            "  \"keypoolsize_hd_internal\": xxxx, (numeric) how many new keys are pre-generated for internal use (change outputs, 10% of the -keypoolsize target, only appears if HD is enabled)\n"
```
I think the 10% number is wrong? current code is 20, I believe? (and I think it should be changed to 100).

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2315,17 +2316,20 @@ UniValue getwalletinfo(const JSONRPCRequest& request)
     obj.push_back(Pair("unconfirmed_balance", ValueFromAmount(pwalletMain->GetUnconfirmedBalance())));
     obj.push_back(Pair("immature_balance",    ValueFromAmount(pwalletMain->GetImmatureBalance())));
     obj.push_back(Pair("txcount",       (int)pwalletMain->mapWallet.size()));
     obj.push_back(Pair("keypoololdest", pwalletMain->GetOldestKeyPoolTime()));
```
I think we really need to update this - keypoololdest is used to know when your backup has been invalidated. I know technically now that we have HD wallets arent invalidated, but until we have rebuilding-from-chain implemented I think we really need to make sure keypoololdest is still useable as "if your backup is older than this date, backup again".

This means keypoololdest needs to be max(oldest internal keypool, oldest external keypool) isntead.

jonasschnelli replied:
Agree. But probably in a separate PR.

TheBlueMatt replied:
Not doing this would break some uses of walletbackup, I think, so I wouldnt be too happy with this slipping into a different release.

ryanofsky replied:
Maybe add a TODO comment (though this does seem to me like a desirable change to include in this PR).

jonasschnelli replied:
> Maybe add a TODO comment (though this does seem to me like a desirable change to include in this PR).

This is fixed in the current version.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2296,16 +2296,17 @@ UniValue getwalletinfo(const JSONRPCRequest& request)
+            "  \"immature_balance\": xxxxxx,      (numeric) the total immature balance of the wallet in " + CURRENCY_UNIT + "\n"
+            "  \"txcount\": xxxxxxx,              (numeric) the total number of transactions in the wallet\n"
+            "  \"keypoololdest\": xxxxxx,         (numeric) the timestamp (seconds since Unix epoch) of the oldest pre-generated key in the key pool\n"
+            "  \"keypoolsize\": xxxx,             (numeric) how many new keys are pre-generated\n"
```
Maybe also mention that keypoolsize_hd_internal is NOT included in this count, as one might assume that external_keypool = keypoolsize - keypoolsize_hd_internal from these docs.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -2929,14 +2961,24 @@ void CWallet::ReserveKeyFromKeyPool(int64_t& nIndex, CKeyPool& keypool)
+                throw runtime_error(std::string(__func__) + ": read failed");
+            if (!HaveKey(tmpKeypool.vchPubKey.GetID()))
+                throw runtime_error(std::string(__func__) + ": unknown key in key pool");
+            if (!IsHDEnabled() || tmpKeypool.fInternal == internal)
```
You need to check if internal is enabled here, not just hd enabled, I think. GetReservedKey is used in a few places with internal set to true without checking if internal is enabled, and if its not, you could end up never succeeding in getting a key, since HD might be enabled, but they never find an internal key.

TheBlueMatt commented on src/wallet/wallet.h:
```diff
@@ -111,6 +116,13 @@ class CKeyPool
             READWRITE(nVersion);
         READWRITE(nTime);
         READWRITE(vchPubKey);
+        if (nVersion >= FEATURE_HD_SPLIT)
```
I do not believe the serializer version is set to the wallet's version. I believe it is the one set in src/wallet/db.h - CLIENT_VERSION.

jonasschnelli committed:
Immediately return setKeyPool's size if HD or HD_SPLIT is disabled or not supported

jonasschnelli committed:
Removed redundant IsLocked() check in NewKeyPool()

jonasschnelli committed:
Fix wrong keypool internal size in RPC getwalletinfo help

jonasschnelli committed:
Make sure ReserveKeyFromKeyPool only hands out internal keys if HD_SPLIT is supported

jonasschnelli committed:
Make sure we hand out keypool keys if HD_SPLIT is not enabled

jonasschnelli committed:
Fix issue where CDataStream->nVersion was taken a CKeyPool record version

jonasschnelli committed:
Make sure we always generate one keypool key at minimum

jonasschnelli commented:
Added a couple of fixes:
* Most important, as @TheBlueMatt has discovered, the `nVersion` handling of `CKeyPool` was wrong. I have fixed it with a `try{}` to serialise the `fInternal` boolean.
* Return `setKeyPool.size()` if hd or hd-chain-split is disabled
* Removed redundant `IsLocked()` check
* Fixed wrong keypool internal size in RPC help
* Make sure we hand out keypool keys if HD_SPLIT is not enabled


I did not cover design-changes and non-important nits in order to get this merged today before the feature freeze. We can fix things and make it more elegant after a possible merge.
I also not changed the keypool distribution (still 100%/20%). Can be done after the freeze and once we all agree whether we'd like to have 100/20 or 100/100.

btcdrak commented:
ACK e34676e

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2296,16 +2296,17 @@ UniValue getwalletinfo(const JSONRPCRequest& request)
+            "  \"immature_balance\": xxxxxx,      (numeric) the total immature balance of the wallet in " + CURRENCY_UNIT + "\n"
+            "  \"txcount\": xxxxxxx,              (numeric) the total number of transactions in the wallet\n"
+            "  \"keypoololdest\": xxxxxx,         (numeric) the timestamp (seconds since Unix epoch) of the oldest pre-generated key in the key pool\n"
+            "  \"keypoolsize\": xxxx,             (numeric) how many new keys are pre-generated (excluding internal-/chnage-output keys)\n"
```
s/chnage/change/

TheBlueMatt replied:
Also, these docs are super confusing - the way I read it, if I have HD enabled I'll never have a keypool for change-output keys, which isnt true - they're just included here.

jonasschnelli committed:
Switch to 100% for the HD internal keypool size

jonasschnelli committed:
Don't switch to HD-chain-split during wallet encryption of non HD-chain-split wallets

jonasschnelli committed:
GetOldestKeyPoolTime: if HD & HD Chain Split is enabled, response max(oldest-internal-key, oldest-external-key)

jonasschnelli commented:
Added three fixes for issues found by @TheBlueMatt and @luke-jr :

* Switched to 100%/100% for the keypool distribution, also, make sure we always generate at least 1 key for each chain
* Added a fix that ensures that we don't accidentally upgrade a NON hd-chain-split wallet to hd-chain-split wallet during encryption (=HD reseed).
* Fix `GetOldestKeyPoolTime` to response `max(oldest_internal_key, oldest_external_key)`.

jonasschnelli committed:
Only show keypoolsize_hd_internal if HD split is enabled

jonasschnelli committed:
CKeyPool avoid "catch (...)" in SerializationOp

TheBlueMatt commented:
utACK a874b75a180a9e967a3d936cf46df11f8531b70a modulo performance concerns, but I'm more than happy to fix those up post-merge if needed, as they should be pretty easy to address

jonasschnelli commented:
Performance on my Core i7 2.9 GHZ with a 5'000 keypool wallet.

This PR (results actually in 5k internal and 5k external keys = 10k keys)
`getnewaddress: real    0m0.219s`

0.13 (5k keys, no chain split)
`getnewaddress: real    0m0.104s`

luke-jr reviewed (approved):
utACK. It's not perfect, but seems to be good enough.

btcdrak commented:
utACK a874b75

gmaxwell commented on src/wallet/wallet.h:
```diff
@@ -88,6 +88,10 @@ enum WalletFeature
 
     FEATURE_HD = 130000, // Hierarchical key derivation after BIP32 (HD Wallet)
+
+    //TODO: FEATURE_HD_SPLIT needs to be bumped to 140000 once we branch of 0.14 //
```
I'm completely confused by this? why would the version of the wallet change due to a branch here? won't that break compatibility with wallets generated by git master and cause funds loss for them?

jonasschnelli replied:
We can keep `139999` for 0.14 to make it compatible with wallets created before branching of 0.14 made with this PR. Unsure what's best practice here...

jonasschnelli replied:
from IRC
```
[21:24:20]  <sipa>	[20:47:45] yeah, there is a race there
[21:24:20]  <sipa>	[20:48:00] that's always the case when a feature needs to be tied to a version number
[21:24:20]  <morcos>	[20:48:18] i don't see any problem leaving FEATURE_HD_SPLIT = 139900
[21:24:20]  <jonasschnelli>	[20:48:21] sipa: keeping in 139999 looks bad but is efficient?
[21:24:20]  <morcos>	[20:48:24] that seems the correct way to do it
[21:24:20]  <jonasschnelli>	[20:48:41] Agree with morcos 
[21:24:20]  <sipa>	[20:48:54] a better way would be to disentangle the wallet version number for the software version number
[21:24:20]  <sipa>	[20:49:09] so the wallet version can just be bumped in the same PR as the feature is introduced
```

ryanofsky replied:
Sounds like you just need to remove the TODO.

jonasschnelli replied:
I think if we branch of 0.14 we would need to bump it to `1499000`. If someone was using this patch in production... hes.. in serious troubles anyways. :)

MarcoFalke replied:
It should not be required to bump it when it is in master, because some people are using master "in production" (i.e. on the main net).

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3027,9 +3027,27 @@ int64_t CWallet::GetOldestKeyPoolTime()
+        }
+        return std::max(oldest_internal, oldest_external);
+    }
+    // load oldest key from keypool, get time and return
```
Could you add a comment mentioning why in this case it is safe to assume key with the lowest index is the oldest key, while in the above case, you need to loop through the keys to find the oldest one. It's not clear (at least to me) why FEATURE_HD_SPLIT would affect key order.

jonasschnelli replied:
I haven't change this part,... but I feel with you that this is very likely a bug we carry around a long time. Haven't verified it, but returning a `CKeyPool` done in (`void CWallet::ReturnKey(int64_t nIndex)`) will mess up the order assumption.
If is a bug, we should fix it,... but maybe in a different PR.

jonasschnelli replied:
I debugged that a little bit and it seems that this assumption (first item in setKeyPool is always oldest key) holds.

* The `std::set setKeyPool` (always ordered) contains `int64_t`. CKeyPool instances remember that nIndex (index of the setKeyPool int64.
* New CKeyPools will always insert `int64_t nEnd = *(--setKeyPool.end()) + 1;`.
* Returning keys will keep the date order because the returned will be inserted at the position where it has been taken out.

The only problem I can see is when someone overflows int64 by creating 2^63+1 keys. But, well,... yes.

ryanofsky replied:
So it seems like there is no reason for having a FEATURE_HD_SPLIT special case here. Could your remove either the new block of code or the old block of code, and just keep one implementation?

jonasschnelli replied:
We still need the `FEATURE_HD_SPLIT` special case here. If don't want to return just the oldest key in that case. 
Reason:
* Assume people use GetOldestKeyPoolTime() to determine if it's time for a new backup (difficult question when it's HD anyways, but that actually what @TheBlueMatt came up in https://github.com/bitcoin/bitcoin/pull/9294/files#r96107555)
* Assume your key in the keypool at index 0 (=oldest key) is an external key.
* Now assume you call `getrawchangeaddress` multiple times and you only and completely refill your internal keypool.
* In that case, you want to return the oldest internal key because its very likely newer then your oldest external key.

But maybe I'm wrong.

ryanofsky replied:
I think you're right. Actually I was just confused because I misread std::max as std::min and couldn't figure out what the new code was doing differently!

ryanofsky reviewed (commented):
utACK a874b75a180a9e967a3d936cf46df11f8531b70a. Suggested some minor changes, probably for a future PR since this needs to be merged soon.

jonasschnelli commented:
Removed 0.14 tag (added to 0.15 milestone). Let's hope we can get this in after the 0.14 branch-off

ryanofsky reviewed (approved):
a874b75a180a9e967a3d936cf46df11f8531b70a still looks good, though it would be nice if some of the outstanding comments here could be addressed in this PR or a followup.

jonasschnelli committed:
Fix superfluous cast and code style nits in RPC wallet-hd.py test

jonasschnelli committed:
Overhaul the internal/external key derive switch

jonasschnelli committed:
Remove FEATURE_HD_SPLIT bump TODO

jonasschnelli commented:
1. Fixed most (all?) of @ryanofsky nits
2. Fixed @theuni's internal/external design nit.
3. Removed the bump to 140000 TODO

wodry commented:
This split sounds interesting, I wanted to know about the motivation for this, but did not find any information neither in this PR nor by searching in the internet. Could You provide some information/link about motivation for this PR ?

sipa commented:
@wodry BIP32 specifies the use of separate chains for external addresses (those you give out to receive payments on) and internal addresses (used for sending change back to yourself). The initial BIP32 implementation in Bitcoin Core 0.13 did not implement this, and only had a single chain for both. That means you're unable to distinguish what are payments and change in a potential future version that can autonomously recover from a wallet backup.

jonasschnelli commented:
Had to rebase after #9377 and slightly modify the `fundrawtransaction.py` test (change `getnewaddress()` into `getrawchangeaddress()`.

jonasschnelli committed:
Slightly modify fundrawtransaction.py test (change getnewaddress() into getrawchangeaddress())

jonasschnelli commented:
Is anything holding back this PR, is there anything left to do?

jonasschnelli commented:
Needed rebase (caused by #8775).

laanwj commented:
> Is anything holding back this PR, is there anything left to do?

Eh yes we should merge this soon.
Sorry, (again) needs rebase though, probably trivial for the std:: changes in #9643

jonasschnelli commented:
Rebased.

laanwj commented:
Locally here (Ubuntu 16.04, x86_64) it fails the wallet-dump.py test:
```
wallet-dump.py:
2017-03-10 09:49:52.887000 TestFramework (INFO): Initializing test directory /tmp/testb91k5t1l/479
2017-03-10 09:49:55.124000 TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/store/orion/projects/bitcoin/bitcoin/qa/rpc-tests/test_framework/test_framework.py", line 147, in main
    self.run_test()
  File "/home/orion/projects/bitcoin/bitcoin/qa/rpc-tests/wallet-dump.py", line 91, in run_test
    assert_equal(found_addr_rsv, 90*2) # 90 keys plus 100% internal keys
  File "/store/orion/projects/bitcoin/bitcoin/qa/rpc-tests/test_framework/util.py", line 476, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(90 == 180)
2017-03-10 09:49:55.125000 TestFramework (INFO): Stopping nodes
2017-03-10 09:49:57.249000 TestFramework (WARNING): Not cleaning up dir /tmp/testb91k5t1l/479
2017-03-10 09:49:57.250000 TestFramework (ERROR): Test failed. Test logging available at /tmp/testb91k5t1l/479/test_framework.log

Pass: False, Duration: 4 s
```

NicolasDorier commented:
utACK 52874f378d3ba3050fddd3a3af0c5347077e2893

I do not really like the O(N) disk hit on the number of key in the pool. As I am dealing with lots of keys, and machines with crappy latency. But I might fix that in a separate PR.

NicolasDorier commented:
Removing my ACK for now, this would make ReserveKey really too slow. I will propose a commit to fix that this week.

NicolasDorier commented on src/wallet/wallet.cpp:
```diff
@@ -2915,30 +2932,44 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
+        for (int64_t i = missingInternal + missingExternal; i--;)
         {
             int64_t nEnd = 1;
+            if (i < missingInternal)
```
if you use `internal = `i % 2 == 0`, then at least `ReserveKeyFromPool` would not have to iterates over all the external keys to find one which match the filter.

NicolasDorier replied:
Actually no, internal my answer would not do. But you can still not clutter internal and external together.

jonasschnelli commented:
Needed rebase.

I think the performance is acceptable (see comparison further down).
Improving the performance could be done after this PR. It's already packed.

A `keypool=5000` comparison between this PR and Master (2.9 GHz Intel Core i7, OSX, SSD).

## This PR
(Keep in mind that they keypool is now **x2, 5000 ext. and 5000 int. key**)
```
$ time ./src/bitcoin-cli --regtest --datadir=../dummy_intkp/ getnewaddress
n3yNEDDgo1k512LXBF8k9h2cSCfvsqKPac

real	0m0.204s
user	0m0.003s
sys	0m0.003s

$ time ./src/bitcoin-cli --regtest --datadir=../dummy_intkp/ getrawchangeaddress
mnVRynjpdTtLovncFDWPn2dkqrLnij5Kf9

real	0m0.271s
user	0m0.004s
sys	0m0.003s

```

## Master
```
$ time ./src/bitcoin-cli --regtest --datadir=../dummy_master/ getnewaddress
mmzkFkya299T4ebushyoHye48c5xJmd7GC

real	0m0.089s
user	0m0.004s
sys	0m0.004s

$ time ./src/bitcoin-cli --regtest --datadir=../dummy_master/ getrawchangeaddress
mwZbWTps1EgZoJoF59XR8meWQUA3rNdb2i

real	0m0.071s
user	0m0.003s
sys	0m0.003s
```

NicolasDorier commented:
@jonasschnelli well this is on SSD. My hosted VM does not have that. Can you at least do something where I commented on the `if (i < missingInternal)` ?
You clutter internal and external together which make a very bad worst case scenario when there is a scan of the pool based on the filter.

jonasschnelli commented:
@NicolasDorier: this PR is open since a couple of month (dec 2016) and I don't want to add any form of optimisations (only fix real problems and it seems to me that the performance is still acceptable/while not good). We really should do this after this has been merged.
Adding more optimisation will cause another round of reviewing (and we already did a couple of them).

ryanofsky commented:
utACK 78f2cb926495114354d835a2ba344558384c89f9

Confirmed there were no significant changes in the first 13 commits which were rebased since my previous ACK. The rebased commits just added std:: namespace prefixes and updated various python rpc error checks.

All of the changes in the 4 new commits also looked good. The only one of these that was a change in behavior rather than simple cleanup was the change to fundrawtransaction.py, which seemed fine (though it would be good if the commit message mentioned the reason for the change).

NicolasDorier commented:
utACK 78f2cb9

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -3016,9 +3057,27 @@ int64_t CWallet::GetOldestKeyPoolTime()
+            if (keypool.fInternal && keypool.nTime < oldest_internal)
+                oldest_internal = keypool.nTime;
+            else if (!keypool.fInternal && keypool.nTime < oldest_external)
+                oldest_external = keypool.nTime;
```
Maybe add an additional if (oldest_internal != now && oldest_external != now) break; ?

jonasschnelli replied:
Ah. Right. It's guaranteed that the `setKeyPool` iterates always from the oldest to the newest.
Will add.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -3626,9 +3685,12 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
             if (!walletInstance->SetHDMasterKey(masterPubKey))
                 throw std::runtime_error(std::string(__func__) + ": Storing master key failed");
+
+            // ensure this wallet.dat can only be opened by clients supporting HD with chain split
```
Technically this should happen before SetHDMasterKey, I believe (otherwise it maybe a race?)

TheBlueMatt commented on src/wallet/walletdb.h:
```diff
@@ -46,9 +46,12 @@ class CHDChain
-    static const int CURRENT_VERSION = 1;
+    static const int VERSION_HD_BASE        = 1;
+    static const int VERSION_HD_CHAIN_SPLIT = 2;
+    static const int CURRENT_VERSION        = VERSION_HD_CHAIN_SPLIT;
```
By defaulting to VERSION_HD_CHAIN_SPLIT, encrypting a wallet will auto-upgrade the wallet without asking the user, I believe this is technically a bug.

jonasschnelli replied:
It would upgrade to the new `CHDChain` serialisation format (including the `nInternalChainCounter`), but it would not upgrade to use internal keys, right?
Is that problematic? If so, I can add a fix that keeps the serialisation version of `CHDChain` during encryption.

TheBlueMatt replied:
It would break opening in old wallets either way - because the nInternalChainCounter is added in the middle instead of the end the deserialization would come back with garbage.

TheBlueMatt commented on qa/rpc-tests/fundrawtransaction.py:
```diff
@@ -476,6 +477,7 @@ def run_test(self):
 
         #refill the keypool
         self.nodes[1].walletpassphrase("test", 100)
+        self.nodes[1].keypoolrefill(8) #need to refill the keypool to get an internal change address
```
Why is this required? Doesnt walletpassphrase TopUpKeyPool?

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -137,22 +137,27 @@ void CWallet::DeriveNewChildKey(CKeyMetadata& metadata, CKey& secret)
-    // derive m/0'/0'
-    accountKey.Derive(externalChainChildKey, BIP32_HARDENED_KEY_LIMIT);
+    // derive m/0'/0' (external chain) OR m/0'/1' (internal chain)
+    accountKey.Derive(chainChildKey, BIP32_HARDENED_KEY_LIMIT+(internal ? 1 : 0));
```
Maybe assert(CanSupportFeature()) here?

TheBlueMatt reviewed (commented):
Didn't look too closely at tests

TheBlueMatt commented:
Needs rebase.

jonasschnelli committed:
Make sure we set the wallets min version to FEATURE_HD_SPLIT at the very first point

jonasschnelli committed:
Optimize GetOldestKeyPoolTime(), return as soon as we have both oldest keys

jonasschnelli committed:
Define CWallet::DeriveNewChildKey() as private

jonasschnelli committed:
Add assertion for CanSupportFeature(FEATURE_HD_SPLIT)

jonasschnelli commented:
Addressed @TheBlueMatt's nits (except https://github.com/bitcoin/bitcoin/pull/9294#discussion_r107040500 which I'm not sure if this is a concern or not)

NicolasDorier commented:
re utACK

jonasschnelli committed:
Do not break backward compatibility during wallet encryption

jonasschnelli commented:
Added a commit to address the backward compatibility issue during wallet encryption mentioned by @TheBlueMatt (https://github.com/bitcoin/bitcoin/pull/9294#discussion_r107040500).
@TheBlueMatt: can you give it a re-test?

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2437,18 +2438,23 @@ UniValue getwalletinfo(const JSONRPCRequest& request)
 
     UniValue obj(UniValue::VOBJ);
+
+    size_t kpExternalSize = pwalletMain->KeypoolCountExternalKeys();
```
Oops, should be pwallet, not pwalletMain now :).

TheBlueMatt commented on src/wallet/walletdb.h:
```diff
@@ -58,13 +61,16 @@ class CHDChain
         READWRITE(this->nVersion);
         READWRITE(nExternalChainCounter);
+        if (this->nVersion >= VERSION_HD_CHAIN_SPLIT)
+            READWRITE(nInternalChainCounter);
```
In addition to the fix for backward compat where you set the version, can we make this more robust and flip the order? Generally adding new fields for (de-)serialization we should be adding them to the end.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -633,7 +639,9 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
-            if (!SetHDMasterKey(masterPubKey))
+            // preserve the old chains version to not break backward compatibility
+            CHDChain oldChain = GetHDChain();
+            if (!SetHDMasterKey(masterPubKey, &oldChain))
```
Why not SetHDMasterKey(int nHDChainVersion)? Then you can just do CanSupportFeature(FEATURE_HD_SPLIT) ? CHDChain::VERSION_HD_CHAIN_SPLIT : CHDChain::VERSION_HD_BASE.

jonasschnelli replied:
I though hiding the internals at this point would be preferable. IMO `CWallet::encryptWallet` should not have direct knowhow of the CHDChain version handling. I'd prefer passing in the old chain and let it be handled through `SetHDMasterKey`. Also, in future maybe other attributes need to be preserved.
Not sure although.

TheBlueMatt replied:
It kinda feels like mixing models, then. Wallet checks for FEATURE_HD_SPLIT everywhere before it updates the nInternalCounts, ie it seems to think its responsible for doing the checks, only to have CHDChain hide some stuff like version, but then you have to copy the version? Either we move logic into CHDChain (as @luke-jr suggested previously, have CHDChain know what version means and assert if you try to use the internal counter with a version that doesnt support it, which I assume some folks would object to), or all the logic should be in CWallet.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -3017,9 +3065,33 @@ int64_t CWallet::GetOldestKeyPoolTime()
+        for(const int64_t& id : setKeyPool)
+        {
+            if (!walletdb.ReadPool(id, keypool)) {
+                throw std::runtime_error(std::string(__func__) + ": read failed");
```
nit: the old version does an assert(keypool.vchPubKey.IsValid()); which is nice to have...not necessary to add, but would be cool.

jonasschnelli replied:
For the keypool oldest it's further down L3026 (at https://github.com/bitcoin/bitcoin/pull/9294/files#diff-b2bb174788c7409b671c46ccc86034bdL3026)

jonasschnelli committed:
Fix rebase issue where pwalletMain was used instead of pwallet
Ser./Deser. nInternalChainCounter as last element

jonasschnelli commented:
Fixed two points reported by @TheBlueMatt:
1. Move Ser./Deser. of `nInternalChainCounter` to be the last element
2. Fix a rebase issue where `pwalletMain` was used instead of `pWallet`.

laanwj commented:
utACK 4115af7
I think this has enough ACKs. Let's merge this and fix the rest, which seems to be less critical, later.
